---
title: "Class13"
author: "Michelle Janossy"
date: "5/14/2019"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Sample genotypes in the MXL 1000 genome project data

```{r}
#Read csv from ensemble
mxl <- read.csv("373531-SampleGenotypes-Homo_sapiens_Variation_Sample_rs8067378.csv")
head(mxl)
```

How many of each genotype are there?

```{r}
table(mxl$Genotype..forward.strand.)
```

Proportion or percent of total for each genotype

```{r}
(table(mxl$Genotype..forward.strand.)/nrow(mxl))*100
```

## Quality Scores in FASTQ files

The forth line of a FASTQ sequence format file encodes the quality score that tells us how good the sequence at a given position is (i.e. how likely it is to be correct based in the instrument)

```{r}
library(seqinr)
library(gtools)
```

## Population Scale analysis

Read RNA-seq count data with genotype information results table

```{r}
expr <- read.table("rs8067378_ENSG00000172057.6.txt")
head(expr)
```

```{r}
summary(expr)
```

```{r}
inds <- expr$geno == "G/G"
summary(expr[inds, "exp"])
```

```{r}
inds <- expr$geno == "A/G"
summary(expr[inds, "exp"])
```

```{r}
inds <- expr$geno == "A/A"
summary(expr[inds, "exp"])
```

```{r}
boxplot(exp ~ geno , data=expr)
```










